File: src/spatial_ffi.erl
-module(spatial_ffi).
-export([box_intersects/4, box_contains_point/3, sphere_intersects/4,
         distance_squared/2, distance/2, compute_bounds/1, merge_bounds/4]).

%% Fast AABB (Box-Box) intersection test
%% Returns true if two axis-aligned bounding boxes intersect
box_intersects({vec3, MinAX, MinAY, MinAZ}, {vec3, MaxAX, MaxAY, MaxAZ},
               {vec3, MinBX, MinBY, MinBZ}, {vec3, MaxBX, MaxBY, MaxBZ}) ->
    (MinAX =< MaxBX) andalso (MaxAX >= MinBX) andalso
    (MinAY =< MaxBY) andalso (MaxAY >= MinBY) andalso
    (MinAZ =< MaxBZ) andalso (MaxAZ >= MinBZ).

%% Fast box contains point test
%% Returns true if a point is inside an axis-aligned bounding box
box_contains_point({vec3, MinX, MinY, MinZ}, {vec3, MaxX, MaxY, MaxZ},
                   {vec3, PX, PY, PZ}) ->
    (PX >= MinX) andalso (PX =< MaxX) andalso
    (PY >= MinY) andalso (PY =< MaxY) andalso
    (PZ >= MinZ) andalso (PZ =< MaxZ).

%% Fast sphere-sphere intersection test
sphere_intersects({vec3, CenterAX, CenterAY, CenterAZ}, RadiusA,
                  {vec3, CenterBX, CenterBY, CenterBZ}, RadiusB) ->
    DX = CenterAX - CenterBX,
    DY = CenterAY - CenterBY,
    DZ = CenterAZ - CenterBZ,
    DistSq = DX * DX + DY * DY + DZ * DZ,
    RadiusSum = RadiusA + RadiusB,
    DistSq =< RadiusSum * RadiusSum.

%% Fast distance squared calculation
distance_squared({vec3, AX, AY, AZ}, {vec3, BX, BY, BZ}) ->
    DX = AX - BX,
    DY = AY - BY,
    DZ = AZ - BZ,
    DX * DX + DY * DY + DZ * DZ.

%% Fast distance calculation
distance({vec3, AX, AY, AZ}, {vec3, BX, BY, BZ}) ->
    DX = AX - BX,
    DY = AY - BY,
    DZ = AZ - BZ,
    math:sqrt(DX * DX + DY * DY + DZ * DZ).

%% Compute bounding box from list of positions
compute_bounds([]) ->
    %% Return a default large bounding box if empty
    {{vec3, 1.0e10, 1.0e10, 1.0e10}, {vec3, -1.0e10, -1.0e10, -1.0e10}};
compute_bounds([{vec3, X, Y, Z} | Rest]) ->
    compute_bounds_loop(Rest, X, Y, Z, X, Y, Z).

compute_bounds_loop([], MinX, MinY, MinZ, MaxX, MaxY, MaxZ) ->
    {{vec3, MinX, MinY, MinZ}, {vec3, MaxX, MaxY, MaxZ}};
compute_bounds_loop([{vec3, X, Y, Z} | Rest], MinX, MinY, MinZ, MaxX, MaxY, MaxZ) ->
    NewMinX = min(MinX, X),
    NewMinY = min(MinY, Y),
    NewMinZ = min(MinZ, Z),
    NewMaxX = max(MaxX, X),
    NewMaxY = max(MaxY, Y),
    NewMaxZ = max(MaxZ, Z),
    compute_bounds_loop(Rest, NewMinX, NewMinY, NewMinZ, NewMaxX, NewMaxY, NewMaxZ).

%% Merge two bounding boxes
merge_bounds({vec3, MinAX, MinAY, MinAZ}, {vec3, MaxAX, MaxAY, MaxAZ},
             {vec3, MinBX, MinBY, MinBZ}, {vec3, MaxBX, MaxBY, MaxBZ}) ->
    {{vec3, min(MinAX, MinBX), min(MinAY, MinBY), min(MinAZ, MinBZ)},
     {vec3, max(MaxAX, MaxBX), max(MaxAY, MaxBY), max(MaxAZ, MaxBZ)}}.

File: src/spatial_ffi.mjs
// Internal FFI functions for performance-critical operations
// JavaScript implementations for browser/Node.js targets

/**
 * Fast AABB (Box-Box) intersection test
 * Returns true if two axis-aligned bounding boxes intersect
 */
export function box_intersects(min_a, max_a, min_b, max_b) {
  return (
    min_a.x <= max_b.x &&
    max_a.x >= min_b.x &&
    min_a.y <= max_b.y &&
    max_a.y >= min_b.y &&
    min_a.z <= max_b.z &&
    max_a.z >= min_b.z
  );
}

/**
 * Fast box contains point test
 * Returns true if a point is inside an axis-aligned bounding box
 */
export function box_contains_point(min, max, point) {
  return (
    point.x >= min.x &&
    point.x <= max.x &&
    point.y >= min.y &&
    point.y <= max.y &&
    point.z >= min.z &&
    point.z <= max.z
  );
}

/**
 * Fast sphere-sphere intersection test
 */
export function sphere_intersects(center_a, radius_a, center_b, radius_b) {
  const dx = center_a.x - center_b.x;
  const dy = center_a.y - center_b.y;
  const dz = center_a.z - center_b.z;
  const dist_sq = dx * dx + dy * dy + dz * dz;
  const radius_sum = radius_a + radius_b;
  return dist_sq <= radius_sum * radius_sum;
}

/**
 * Fast distance squared calculation
 */
export function distance_squared(a, b) {
  const dx = a.x - b.x;
  const dy = a.y - b.y;
  const dz = a.z - b.z;
  return dx * dx + dy * dy + dz * dz;
}

/**
 * Fast distance calculation
 */
export function distance(a, b) {
  const dx = a.x - b.x;
  const dy = a.y - b.y;
  const dz = a.z - b.z;
  return Math.sqrt(dx * dx + dy * dy + dz * dz);
}

/**
 * Compute bounding box from list of positions
 * Returns [min, max] tuple
 */
export function compute_bounds(positions) {
  // Convert Gleam list to JavaScript array
  const posArray = positions.toArray();

  if (posArray.length === 0) {
    return [
      { x: 1e10, y: 1e10, z: 1e10 },
      { x: -1e10, y: -1e10, z: -1e10 },
    ];
  }

  let min_x = posArray[0].x;
  let min_y = posArray[0].y;
  let min_z = posArray[0].z;
  let max_x = posArray[0].x;
  let max_y = posArray[0].y;
  let max_z = posArray[0].z;

  for (let i = 1; i < posArray.length; i++) {
    const pos = posArray[i];
    if (pos.x < min_x) min_x = pos.x;
    if (pos.y < min_y) min_y = pos.y;
    if (pos.z < min_z) min_z = pos.z;
    if (pos.x > max_x) max_x = pos.x;
    if (pos.y > max_y) max_y = pos.y;
    if (pos.z > max_z) max_z = pos.z;
  }

  return [
    { x: min_x, y: min_y, z: min_z },
    { x: max_x, y: max_y, z: max_z },
  ];
}

/**
 * Merge two bounding boxes
 * Returns [min, max] tuple
 */
export function merge_bounds(min_a, max_a, min_b, max_b) {
  return [
    {
      x: Math.min(min_a.x, min_b.x),
      y: Math.min(min_a.y, min_b.y),
      z: Math.min(min_a.z, min_b.z),
    },
    {
      x: Math.max(max_a.x, max_b.x),
      y: Math.max(max_a.y, max_b.y),
      z: Math.max(max_a.z, max_b.z),
    },
  ];
}
